cfg_mgmt/btp_application_certificate.py

```diff
@@ -0,0 +1,256 @@
+            if cn:
+                serial_no, base = BtpApplicationCertificate.get_serial_no_from_common_name(cn)
+                if serial_no > 0 and common_name_base == base:
+                    certs.append(CertInfo(id=id, dn=dn, cn=cn, serial_no=serial_no))
+        return certs
+
+
+_str_to_names = {
+    'C': NameOID.COUNTRY_NAME,
+    'ST': NameOID.STATE_OR_PROVINCE_NAME,
+    'O': NameOID.ORGANIZATION_NAME,
+    'OU': NameOID.ORGANIZATIONAL_UNIT_NAME,
+    'L': NameOID.LOCALITY_NAME,
+    'CN': NameOID.COMMON_NAME,
+    'EMAIL': NameOID.EMAIL_ADDRESS,
+}
+
+
+def _create_csr(subject: str) -> tuple[str, str]:
+    logger.info(f'Creating CSR for {subject}')
+    key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
+    key_pem = key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.NoEncryption(),
+    ).decode('utf-8')
+    attributes = []
+    for part in subject.split(', '):
+        (name, value) = part.split('=')
+        oid = _str_to_names[name]
+        attributes.append(x509.NameAttribute(oid, value))
+    csr = x509.CertificateSigningRequestBuilder().subject_name(
+        x509.Name(attributes)
+    ).sign(key, hashes.SHA256())
+    csr_pem = csr.public_bytes(serialization.Encoding.PEM).decode('utf-8')
+    return csr_pem, key_pem
+
+
+def _extract_client_certificate(cert_response: dict) -> str:
+    pkcs7_pem = cert_response['value']
+    certs = pkcs7.load_pem_pkcs7_certificates(pkcs7_pem.encode('utf-8'))
+    if len(certs) == 0:
+        raise ValueError('no certificates found in response')
+    return certs[0].public_bytes(serialization.Encoding.PEM).decode('utf-8')
+
+
+def rotate_cfg_element(
+    cfg_element: BtpApplicationCertificate,
+    cfg_factory: model.ConfigFactory,
+) -> typing.Tuple[cfg_mgmt.revert_function, dict, model.NamedModelElement]:
+    gbaas_auth = cfg_factory.btp_application_certificate(cfg_element.auth_application_certificate())
+    gbaas_client = GBaasAppClient(gbaas_auth)
+
+    # calc next serial no
+    cn = cfg_element.common_name()
+    serial_no, base = BtpApplicationCertificate.get_serial_no_from_common_name(cn)
+    if serial_no == 0:
+        raise ValueError(f'unexpected cn: {cn}')
+    next_sn = serial_no + 1
+    for info in gbaas_client.list_certificates_by_base(base):
+        if info.serial_no >= next_sn:
+            next_sn = info.serial_no + 1
+    next_cn = f'{next_sn}.{base}'
+
+    # create certificate
+    csr_pem, key_pem = _create_csr(cfg_element.subject(next_cn))
+    sb_auth = cfg_factory.btp_service_binding(cfg_element.cert_service_binding())
+    cs_client = CertServiceClient(sb_auth.credentials())
+    response = cs_client.get_client_certificate_chain(csr_pem, cfg_element.validity_in_days())
+    cert_pem = _extract_client_certificate(response)
+
+    # add certificate to GBaas application
+    id = gbaas_client.put_certificate(
```

Comment:
the comment says `add`, the method says `put` (the latter implies that the existing value would be replaced). which one is it? :-)

Reply:
The API uses a `PUT` request to add a certificate :-)

Reply:
I see. That's a bad choice of HTTP-Methods on API-side, then (but of course nothing we can influence on our end)